File: bit_manipulation/chalkboard_xor.py

def determine_winner(numbers: list[int]) -> bool:  # LeetCode Q.810.
    xor_value = 0
    numbers2counts, distinct_numbers = dict(), []
    for number in numbers:
        xor_value ^= number
        if number not in numbers2counts.keys():
            distinct_numbers.append(number)
            numbers2counts.update({number: 0})
        numbers2counts[number] += 1

    total_finished_plays = 0  # When total plays are even, it's Alice's turn.
    removed_idx = 0
    while distinct_numbers:
        if xor_value == 0:
            return True if total_finished_plays % 2 == 0 else False

        removed_idx -= removed_idx  # Reset to the default 0.
        if distinct_numbers[0] == xor_value and len(distinct_numbers) > 1:
            removed_idx += 1  # Must and can switch to another removal choice.

        removed_num = distinct_numbers[removed_idx]
        numbers2counts[removed_num] -= 1
        if numbers2counts[removed_num] == 0:
            distinct_numbers.pop(removed_idx)

        xor_value ^= removed_num
        total_finished_plays += 1

    return True if total_finished_plays % 2 == 0 else False

File: bit_manipulation/chalkboard_xor/chalkboard_xor.cpp
#include <vector>
#include <unordered_map>
using namespace std;

bool determine_winner(vector<int> &nums) // LeetCode Q.810.
{
    vector<int> distinct_nums;
    unordered_map<int, int> nums2counts;

    int xor_value = 0;
    for (auto num : nums)
    {
        xor_value ^= num;

        if (nums2counts.find(num) == nums2counts.end())
        {
            nums2counts[num] = 0;
            distinct_nums.push_back(num);
        }
        nums2counts[num] += 1;
    }

    int current_round = 1;
    int pop_idx = 0;
    while (xor_value != 0)
    {
        if (distinct_nums[pop_idx] == xor_value && distinct_nums.size() > 1)
        {
            pop_idx += 1; // Must and can switch to another removal choice.
        }

        int erased_num = distinct_nums[pop_idx];
        xor_value ^= erased_num;
        nums2counts[erased_num] -= 1;

        if (nums2counts[erased_num] == 0)
        {
            nums2counts.erase(erased_num);
            distinct_nums.erase(distinct_nums.begin() + pop_idx);
        }

        current_round += 1;
        pop_idx -= pop_idx; // Reset to the default 0.
    }

    if (current_round % 2 == 0)
    {
        return false;
    }
    return true;
}
